src/display/annotation_layer.js

```diff
@@ -1824,157 +1836,197 @@ class ChoiceWidgetAnnotationElement extends WidgetAnnotationElement {
     this.container.classList.add("popupAnnotation");
 
-    const parentElements = this.layer.querySelectorAll(
-      `[data-annotation-id="${this.data.parentId}"]`
-    );
-    if (parentElements.length === 0) {
-      return this.container;
+    let open = this.data.open;
+    if (typeof PDFJSDev !== "undefined" && PDFJSDev.test("TESTING")) {
+      open = true;
     }
 
-    const popup = new PopupElement({
+    this.popup = new PopupElement({
       container: this.container,
-      trigger: Array.from(parentElements),
       color: this.data.color,
       titleObj: this.data.titleObj,
       modificationDate: this.data.modificationDate,
       contentsObj: this.data.contentsObj,
       richText: this.data.richText,
+      rect: this.data.rect,
+      parentRect: this.data.parentRect || null,
+      parent: this.parent,
+      elements: this.elements,
+      open,
     });
+    const elementIds = [];
+    for (const element of this.elements) {
+      element.popup = this.popup;
+      elementIds.push(element.data.id);
+    }
+    this.container.setAttribute("aria-controls", elementIds.join(","));
 
-    // Position the popup next to the parent annotation's container.
-    // PDF viewers ignore a popup annotation's rectangle.
-    const page = this.page;
-    const rect = Util.normalizeRect([
-      this.data.parentRect[0],
-      page.view[3] - this.data.parentRect[1] + page.view[1],
-      this.data.parentRect[2],
-      page.view[3] - this.data.parentRect[3] + page.view[1],
-    ]);
-    const popupLeft =
-      rect[0] + this.data.parentRect[2] - this.data.parentRect[0];
-    const popupTop = rect[1];
-
-    const { pageWidth, pageHeight, pageX, pageY } = this.viewport.rawDims;
-
-    this.container.style.left = `${(100 * (popupLeft - pageX)) / pageWidth}%`;
-    this.container.style.top = `${(100 * (popupTop - pageY)) / pageHeight}%`;
-
-    this.container.append(popup.render());
     return this.container;
   }
 }
 
 class PopupElement {
-  constructor(parameters) {
-    this.container = parameters.container;
-    this.trigger = parameters.trigger;
-    this.color = parameters.color;
-    this.titleObj = parameters.titleObj;
-    this.modificationDate = parameters.modificationDate;
-    this.contentsObj = parameters.contentsObj;
-    this.richText = parameters.richText;
-    this.hideWrapper = parameters.hideWrapper || false;
-
+  #dateTimePromise = null;
+
+  #boundToggle = this.#toggle.bind(this);
+
+  #boundHide = this.#hide.bind(this);
+
+  #boundShow = this.#show.bind(this);
+
+  #boundMouseClick = this.#mouseClick.bind(this);
+
+  constructor({
+    container,
+    color,
+    elements,
+    titleObj,
+    modificationDate,
+    contentsObj,
+    richText,
+    parent,
+    rect,
+    parentRect,
+    open,
+  }) {
+    this.container = container;
+    this.titleObj = titleObj;
+    this.contentsObj = contentsObj;
+    this.richText = richText;
+    this.parent = parent;
+    this.popup = null;
     this.pinned = false;
```

Comment:
Should we perhaps just change this into a private field now, since this state really ought to be *private*?